cypress/integration/login.spec.js

```diff
@@ -0,0 +1,28 @@
+describe('Login Page', () => {
+  beforeEach(() => {
+    cy.visit('/login');
+  });
+
+  it('displays the login form', () => {
+    cy.get('form').should('be.visible');
+    cy.get('input[type="text"]').should('be.visible');
+    cy.get('input[type="password"]').should('be.visible');
+    cy.get('button[type="submit"]').should('be.visible');
+  });
+
+  it('redirects to the dashboard on successful login', () => {
+    cy.get('input[type="text"]').type('username');
+    cy.get('input[type="password"]').type('password');
+    cy.get('button[type="submit"]').click();
+    cy.url().should('include', '/dashboard');
+  });
+
+  it('displays an error message on failed login', () => {
+    cy.get('input[type="text"]').type('wrongusername');
+    cy.get('input[type="password"]').type('wrongpassword');
+    cy.get('button[type="submit"]').click();
+    cy.on('window:alert', (str) => {
+      expect(str).to.equal('Login failed');
```

Comment:
Listening to window:alert may not be reliable and can lead to flaky tests. Consider checking for an element on the page that displays the error message.